src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs

```diff
@@ -25,64 +24,90 @@ internal partial class SuppressMessageAttributeState
-                    {
-                        _globalSymbolSuppressions[symbol] = suppressions.Add(id);
-                    }
+                    AddOrUpdate(info, suppressions);
                 }
                 else
                 {
-                    _globalSymbolSuppressions.Add(symbol, ImmutableArray.Create(id));
+                    suppressions = new Dictionary<string, SuppressMessageInfo>() {{ info.Id, info }};
+                    _globalSymbolSuppressions.Add(symbol, suppressions);
                 }
             }
 
-            public bool HasCompilationWideSuppression(string id)
+            public bool HasCompilationWideSuppression(string id, out SuppressMessageInfo info)
             {
-                return _compilationWideSuppressions.Contains(id);
+                return _compilationWideSuppressions.TryGetValue(id, out info);
             }
 
-            public bool HasGlobalSymbolSuppression(ISymbol symbol, string id)
+            public bool HasGlobalSymbolSuppression(ISymbol symbol, string id, out SuppressMessageInfo info)
             {
                 Debug.Assert(symbol != null);
-                ImmutableArray<string> suppressions;
-                return _globalSymbolSuppressions.TryGetValue(symbol, out suppressions) && suppressions.Contains(id);
+                Dictionary<string, SuppressMessageInfo> suppressions;
+                if (_globalSymbolSuppressions.TryGetValue(symbol, out suppressions) &&
+                    suppressions.TryGetValue(id, out info))
+                {
+                    return true;
+                }
+
+                info = default(SuppressMessageInfo);
+                return false;
             }
         }
 
-        public SuppressMessageAttributeState(Compilation compilation)
+        internal SuppressMessageAttributeState(Compilation compilation)
         {
             _compilation = compilation;
+            _localSuppressionsBySymbol = new ConcurrentDictionary<ISymbol, ImmutableDictionary<string, SuppressMessageInfo>>();
         }
 
-        public bool IsDiagnosticSuppressed(Diagnostic diagnostic, ISymbol symbolOpt = null)
+        public static Diagnostic ApplySourceSuppressions(Diagnostic diagnostic, Compilation compilation, ISymbol symbolOpt = null)
         {
-            if (symbolOpt != null && IsDiagnosticSuppressed(diagnostic.Id, symbolOpt))
+            if (diagnostic.IsSuppressed)
+            {
+                // Diagnostic already has a source suppression.
+                return diagnostic;
+            }
+
+            var suppressMessageState = AnalyzerDriver.GetOrCreateCachedCompilationData(compilation).SuppressMessageAttributeState;
+
+            SuppressMessageInfo info;
+            if (suppressMessageState.IsDiagnosticSuppressed(diagnostic, out info))
+            {
+                // Attach the suppression info to the diagnostic.
+                diagnostic = diagnostic.WithIsSuppressed(true);
```

Comment:
Yes, the decision was driven on the basis that "Diagnostic.IsSuppressed" equates to a specific instance of a diagnostic getting suppressed and the only way to do so currently is through source suppressions. Other means of suppression that you mention are basically "Disable/Enable" rule operations, done at the compilation level and apply to entire rule, not specific diagnostic instances.